contracts/allocators/AuraAllocatorV2.sol

```diff
@@ -0,0 +1,236 @@
+        if (auraBalBalance > 0) abStaking.stake(auraBalBalance);
+
+        uint256 received = _amountAllocated(tokenIndex);
+        uint256 last = extender.getAllocatorAllocated(id) + extender.getAllocatorPerformance(id).gain;
+
+        if (received >= last) gain = uint128(received - last);
+        else loss = uint128(last - received);
+    }
+
+    function deallocate(uint256[] memory amounts) public override onlyGuardian {
+        if (amounts[0] > 0) locker.processExpiredLocks(false);
+        if (amounts[1] > 0) abStaking.withdraw(amounts[1], true); // does this need to be withdrawAndUnwrap?
+    }
+
+    function _deactivate(bool panic) internal override {
+        uint256[] memory amounts = new uint256[](2);
+
+        if (_unlockable() > 0) amounts[0] = 1;
+        if (abStaking.balanceOf(address(this)) > 0) amounts[1] = abStaking.balanceOf(address(this));
+
+        deallocate(amounts);
+
+        if (panic) {
+            aura.transfer(treasury, aura.balanceOf(address(this)));
+
+            // auraBal should be included in the reward tokens
+            uint256 numRewardTokens = _rewardTokens.length;
+            for (uint256 i; i < numRewardTokens; ) {
+                IERC20 token = _rewardTokens[i];
+                uint256 tokenBalance = token.balanceOf(address(this));
+                if (tokenBalance > 0) token.transfer(treasury, tokenBalance);
+
+                unchecked {
+                    ++i;
+                }
+            }
+        }
+    }
+
+    function _prepareMigration() internal override {
+        uint256[] memory amounts = new uint256[](2);
+        amounts[0] = 1;
+        amounts[1] = abStaking.balanceOf(address(this));
+        deallocate(amounts);
+    }
+
+    function amountAllocated(uint256 id) public view override returns (uint256) {
+        uint256 tokenIndex = tokenIds[id];
+        return _amountAllocated(tokenIndex);
+    }
+
+    function rewardTokens() public view override returns (IERC20[] memory) {
+        IERC20[] memory tokens = new IERC20[](1);
+        tokens[0] = auraBal;
+        return tokens;
+    }
+
+    function utilityTokens() public view override returns (IERC20[] memory) {
+        IERC20[] memory tokens = new IERC20[](1);
+        tokens[0] = auraBal;
+        return tokens;
+    }
+
+    function name() external pure override returns (string memory) {
+        return "AuraAllocatorV2";
+    }
+
+    // ========= ALLOCATOR SPECIFIC FUNCTIONS ========= //
+
+    function addRewardToken(address token_) external onlyGuardian {
+        IERC20 token = IERC20(token_);
+        token.approve(address(extender), type(uint256).max);
+
+        _rewardTokens.push(token);
+    }
+
+    function delegate(address delegatee_) external onlyGuardian {
+        locker.delegate(delegatee_);
+    }
+
+    function setShouldLock(bool shouldLock_) external onlyGuardian {
+        shouldLock = shouldLock_;
+    }
+
+    function setLocker(address locker_) external onlyGuardian {
```

Comment:
Does this function assume that the existing locker doesn't hold anymore funds? If so, you may want to check that before setting new value.